src/metabase/upload.clj

```diff
@@ -74,48 +74,43 @@
 ;;
 ;; </code></pre>
 
-(def ^:private type+parent-pairs
-  ;; listed in depth-first order
-  '([::boolean-or-int ::boolean]
-    [::boolean-or-int ::int]
-    [::auto-incrementing-int-pk ::int]
-    [::int ::float]
-    [::date ::datetime]
-    [::boolean ::varchar-255]
-    [::offset-datetime ::varchar-255]
-    [::datetime ::varchar-255]
-    [::float ::varchar-255]
-    [::varchar-255 ::text]))
-
-(defn ^:private column-type
-  "Returns the type of a column given the lowest common ancestor type of the values in the column."
-  [type]
-  (case type
-    ::boolean-or-int ::boolean
-    type))
-
-(def ^:private type->parents
-  (reduce
-   (fn [m [type parent]]
-     (update m type conj parent))
-   {}
-   type+parent-pairs))
+(def ^:private h
+  (-> (make-hierarchy)
+      (derive ::boolean-or-int ::boolean)
+      (derive ::boolean-or-int ::int)
+      (derive ::auto-incrementing-int-pk ::int)
+      (derive ::int ::float)
+      (derive ::date ::datetime)
+      (derive ::boolean ::varchar-255)
+      (derive ::offset-datetime ::varchar-255)
+      (derive ::datetime ::varchar-255)
+      (derive ::float ::varchar-255)
+      (derive ::varchar-255 ::text)))
 
 (def ^:private value-types
-  "All value types including the root type, ::text"
-  (conj (keys type->parents) ::text))
+  (into (set (keys (:children h))) (keys (:parents h))))
+
+(def ^:private concrete-type?
+  "Nodes which can correspond to a concrete database column type."
+  (complement #{::boolean-or-int}))
 
 (def ^:private column-types
   "All column types"
-  (map column-type value-types))
+  (into #{} (filter concrete-type?) value-types))
+
+(defn- column-type [tag]
+  (if (concrete-type? tag)
+    tag
+    (case type
+      ::boolean-or-int ::boolean)))
 
 (defn- bfs-ancestors [type]
-  (loop [visit   (list type)
+  (loop [up-next (list type)
          visited (ordered-set/ordered-set)]
-    (if (empty? visit)
+    (if (empty? up-next)
       visited
-      (let [parents (mapcat type->parents visit)]
-        (recur parents (into visited parents))))))
+      (let [next-types (mapcat parents up-next)]
```

Comment:
Renamed the binding to avoid punning on `clojure.core/parents` in the same line.